File: 백준/Gold/14503. 로봇 청소기/로봇 청소기.py
import sys

input_data = [list(map(int, line.split())) for line in sys.stdin]
n, m = input_data[0]
r, c, d = input_data[1]
room = input_data[2:]

# 북, 동, 남, 서
dr = [-1, 0, 1, 0]
dc = [0, 1, 0, -1]

cleaned = [[False] * m for _ in range(n)]
answer = 0

while True:
    if not cleaned[r][c]:
        cleaned[r][c] = True
        answer += 1
    
    is_dirty_exist = False
    for _ in range(4):
        d = (d - 1) % 4
        nr, nc = r + dr[d], c + dc[d]
        if 0 <= nr < n and 0 <= nc < m and room[nr][nc] == 0 and not cleaned[nr][nc]:
            r, c = nr, nc
            is_dirty_exist = True
            break
    
    if not is_dirty_exist:
        br, bc = r - dr[d], c - dc[d]
        if br < 0 or br >= n or bc < 0 or bc >= m or room[br][bc] == 1:
            break
        r, c = br, bc
    
print(answer)

File: 백준/Gold/14503. 로봇 청소기/로봇 청소기.js
const fs = require('fs');
const input = fs.readFileSync(0).toString().trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);
let [r, c, d] = input[1].split(' ').map(Number);
const room = input.slice(2).map(line => line.split(' ').map(Number));

const dr = [-1, 0, 1, 0];
const dc = [0, 1, 0, -1];

const cleaned = Array.from({ length: n }, () => Array(m).fill(false));
let answer = 0;

while (true) {
    if (!cleaned[r][c]) {
        cleaned[r][c] = true;
        answer++;
    }

    let isDirtyExist = false;

    for (let i = 0; i < 4; i++) {
        d = (d + 3) % 4; // 왼쪽으로 회전
        const [nr, nc] = [r + dr[d], c + dc[d]];

        if (nr >= 0 && nr < n && nc >= 0 && nc < m && room[nr][nc] === 0 && !cleaned[nr][nc]) {
            [r, c] = [nr, nc];
            isDirtyExist = true;
            break;
        }
    }

    if (!isDirtyExist) {
        const [br, bc] = [r - dr[d], c - dc[d]];

        if (br < 0 || br >= n || bc < 0 || bc >= m || room[br][bc] === 1) {
            break;
        }
        [r, c] = [br, bc];
    }
}

console.log(answer);